Support custom controllers as pod owners

irfanurrehman opened the pull request:
**Background**
There are instances where a pod is created, owned and/or controlled by a controller which is not one of the built in k8s workload controllers. Further there could be pods managed and/or updated by multiple controllers but owned by 1. In such cases the controllerRef (`ownerRef.controller: true`) was not set. For such cases we used to skip discovering the pod and creating a workload controller.

**Implementation**
- Update the logic to not exclude the ownership data if the `ownerRef.Controller` is `false`, if it is the only `ownerRef` in the list.
- If there are multiple owners in the list, select the one with `ownerRef.Controller` `true`.
- If there are multiple owners in the list, but none with `ownerRef.Controller` set as `true`. Select the first owner in the list.

**Testing**

The pods with customer controllers and `controllerRef` unset are discovered and their controllers are discovered as customer controllers:

All pods discovered in the namespace
![image](https://user-images.githubusercontent.com/10027921/143581215-9d58158e-474e-46db-932a-00105af3a5dd.png)
![image](https://user-images.githubusercontent.com/10027921/143581453-8ca5130d-9a93-4597-aba6-5f5e4c8edb83.png)


Workload controller discovered as custom controller
![image](https://user-images.githubusercontent.com/10027921/143581410-e3eb01ad-23b8-475c-b07f-cf5683032b68.png)

enlinxu commented on pkg/discovery/util/owner_util.go:
```diff
@@ -19,13 +19,16 @@ type OwnerInfo struct {
 func GetOwnerInfo(owners []metav1.OwnerReference) (OwnerInfo, bool) {
```
We may want to refactor the logic here a bit. Now we removed the continue statement when the IsController is false. In this case, the ownerSet var will always return true unless the owners is empty. 
Also, this logging below will be confusing too,
if ownerSet {
				glog.V(3).Infof("Multiple controller owners found: %+v", owner)
				continue
			}
In the case when you have 3 owners , the 2nd one is the controller one, and when you iterate to the 2nd one, ownerSet is set to true already and the log will be printed, which will be confused.

irfanurrehman replied:
Thanks @enlinxu for the diligence. Updated and improved the logging too.

enlinxu commented on pkg/discovery/util/owner_util.go:
```diff
@@ -19,28 +19,46 @@ type OwnerInfo struct {
+
 		if len(owner.Kind) > 0 && len(owner.Name) > 0 && len(owner.UID) > 0 {
-			// Found a valid controller owner
+			if controllerIndex == i {
```
Why can't we move the IsController(owner) check within the if len(owner.Kind) > 0 && len(owner.Name) > 0 && len(owner.UID) > 0 , so that we don't need the controllerIndex at all.

irfanurrehman replied:
Updated!

irfanurrehman commented:
I did go ahead and update this PR to disable actions on the pods and containers of custom controllers.
@enlinxu @evat-pm please let me know if this needs to be done in a different way. For example, I could think of handling this by still generating actions but make them not executable, with enough information for the user to know why this action is not executable. 
 
I updated this because if we leave it as is, we will generate actions on such pods and containers, but when executed these will fail. We cannot have a generic way to execute such actions. Some controllers might have template specs or fields which reflect into the pods, and some don't. Infact an operator managing pods directly is also an example of a custom controller.

I have updated the fix in the second commit on this PR.
**Testing**
Before the fix we see actions (if I have a pod with enough cpu or mem usage, or no usage for a long time):
![image](https://user-images.githubusercontent.com/10027921/144043813-64081ed6-2571-4023-8e3d-0c35a4bb12c5.png)

![image](https://user-images.githubusercontent.com/10027921/144043861-bbc1e577-98d6-416b-9b8d-21adf2c86e47.png)


After the change no actions generated even with high resource usage:
![image](https://user-images.githubusercontent.com/10027921/144044172-2c4dfbf8-d507-4609-aad1-b1e65cbba6e1.png)

irfanurrehman commented:
@enlinxu Updated as what we discussed. The actions will still be generated, but fail on execution. Will initiate a separate effort for a way to handle the action failures.
